libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>
+ *
+ * With modifications for Ardupilot CAN driver
+ * Copyright (C) 2017 Eugene Shamaev
+ */
+
+//#define CONFIG_HAL_BOARD HAL_BOARD_PX4
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/system.h>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+#include <cassert>
+#include <cstring>
+#include "CAN.h"
+
+#include <nuttx/irq.h>
+#include <unistd.h>
+#include <drivers/drv_hrt.h>
+
+#include <assert.h>
+
+#include "Scheduler.h"
+
+/*
+ * FOR INVESTIGATION:
+ * AP_HAL::micros64() was called for monotonic time counter
+ * pavel-kirienko: This will work as long as we don't need to synchronize the autopilot's own clock with an external
+ * time base, e.g. a GNSS time provided by an external GNSS receiver. Libuavcan's STM32 driver supports automatic time
+ * synchronization only if it has a dedicated hardware timer to work with.
+ */
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+
+#define STM32_SYSCLK_FREQUENCY  168000000ul
+#define STM32_HCLK_FREQUENCY    STM32_SYSCLK_FREQUENCY
+#define STM32_PCLK1_FREQUENCY   (STM32_HCLK_FREQUENCY/4)
+
+#ifndef UAVCAN_NULLPTR
+# if UAVCAN_CPP_VERSION >= UAVCAN_CPP11
+#  define UAVCAN_NULLPTR nullptr
+# else
+#  define UAVCAN_NULLPTR NULL
+# endif
+#endif
+
+#include "CAN.h"
+
+extern const AP_HAL::HAL& hal;
+
+#include <AP_UAVCAN/AP_UAVCAN.h>
+
+extern "C" {
+    static int can1_irq(const int irq, void*);
+    static int can2_irq(const int irq, void*);
+}
+
+using namespace PX4;
+
+uint64_t clock::getUtcUSecFromCanInterrupt()
+{
+    return AP_HAL::micros64();
+}
+
+uavcan::MonotonicTime clock::getMonotonic()
+{
+    uint64_t usec = 0;
+    usec = AP_HAL::micros64();
+
+    return uavcan::MonotonicTime::fromUSec(usec);
+}
+
+BusEvent::BusEvent(PX4CANDriver& can_driver) :
+    _signal(0)
+{
+}
+
+BusEvent::~BusEvent()
+{
+}
+
+bool BusEvent::wait(uavcan::MonotonicDuration duration)
+{
+    struct hrt_call wait_call;
+
+    irqstate_t irs = irqsave();
+    if (_signal) {
+        _signal = 0;
+        irqrestore(irs);
+        return true;
+    }
+
+    sem_init(&_wait_semaphore, 0, 0);
+    irqrestore(irs);
+
+    hrt_call_after(&wait_call, duration.toUSec(), (hrt_callout) signalFromCallOut, this);
+    sem_wait(&_wait_semaphore);
+
+    hrt_cancel(&wait_call);
+
+    irs = irqsave();
+    if (_signal) {
+        _signal = 0;
+        irqrestore(irs);
+
+        return true;
+    }
+    irqrestore(irs);
+
+    return false;
+}
+
+void BusEvent::signalFromCallOut(BusEvent *sem)
+{
+    sem_post(&sem->_wait_semaphore);
+}
+
+void BusEvent::signalFromInterrupt()
+{
+    _signal++;
+    sem_post(&_wait_semaphore);
+}
+
+static void handleTxInterrupt(uint8_t iface_index)
+{
+    if (iface_index < UAVCAN_STM32_NUM_IFACES) {
+        uint64_t utc_usec = clock::getUtcUSecFromCanInterrupt();
+        if (utc_usec > 0) {
+            utc_usec--;
```

Comment:
Does a 1usec matter that much? Also, couldn't the interrupt handling be delayed and this wouldn't be correct anyway?
It doesn't matter much, but I did lose 10 minutes trying to understand what the hell did this do and I guess anyone going through this code in the future won't understand it either.